### PR TITLE
Bug510

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Addressing #510

### CODE REVIEW

**Comment by lostmsu** [comment]
Mostly looks good to me, except the `get_buffer` handling of nested buffers.

File: src/Native/LibTorchSharp/THSNormalization.cpp [hunk 1]
```diff
@@ -202,6 +202,15 @@ Tensor THSNN_BatchNorm1d_get_var(const NNModule module)
     return nullptr;
 }
 
+Tensor THSNN_BatchNorm1d_get_batches(const NNModule module)
+{
+    CATCH(
+        auto v = (*module)->as<torch::nn::BatchNorm1d>()->num_batches_tracked;
+    return v.defined() ? ResultTensor(v) : nullptr;
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSNormalization.cpp:209]: Yes, VS keeps doing this with a macro around two or more lines. Annoying.

File: src/TorchSharp/NN/Normalization/BatchNorm1D.cs [hunk 2]
```diff
@@ -100,7 +102,15 @@ public override Tensor forward(Tensor tensor)
...
-                    ConditionallyRegisterParameter("bias", value);
+                    ConditionallyRegisterBuffer("running_var", value);
+                }
+            }
+
+            public Tensor? num_batches_tracked {
+                get {
+                    var res = THSNN_BatchNorm1d_get_batches(handle);
+                    if (res == IntPtr.Zero) { torch.CheckForErrors(); return null; }
+                    return new Parameter(res);
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Normalization/BatchNorm1D.cs:113]: Copy-and-paste error.

File: src/TorchSharp/NN/Normalization/BatchNorm1D.cs [hunk 3]
```diff
@@ -87,11 +89,11 @@ public override Tensor forward(Tensor tensor)
                 set {
                     THSNN_BatchNorm1d_set_mean(handle, (value is null ? IntPtr.Zero : value.Handle));
                     torch.CheckForErrors();
-                    ConditionallyRegisterParameter("bias", value);
+                    ConditionallyRegisterBuffer("running_mean", value);
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Normalization/BatchNorm1D.cs:92]: Yeah, but I'm not sure that will always be the case. Seems more general to force the caller to specify a name.

File: src/TorchSharp/NN/Module.cs [hunk 4]
```diff
@@ -520,6 +520,25 @@ public virtual bool has_parameter(string target)
...
+                /// <param name="target">The fully-qualified string name of the buffer to look for.</param>
+                /// <returns>The tensor referenced by target</returns>
+                public virtual Tensor get_buffer(string target)
+                {
+                    if (_internal_buffers.TryGetValue(target, out var parameter)) {
+                        return parameter;
+                    }
+                    foreach (var child in named_children().Where(nc => target.StartsWith(nc.name))) {
+                        var prefix = child.name + ".";
+                        var p = child.module.get_buffer(target.Remove(0, prefix.Length));
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:535]: Hmm.

Not sure I see how that specific example would happen, but I do see that the 'where' clause is invalid and may include a number of children that should not be included. The correct code should split the input on '.' and then filter the children on the first element of the resulting array.

Thanks for identifying.

[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs]: My update also has a bug in it. I'll fix that later.

[the remark this answers is not in the document]
Comment (lostmsu) [reply src/TorchSharp/NN/Module.cs]: @NiklasGustafsson do you want to merge this then fix the bug separately?

[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs]: @lostmsu -- "fix it later" as in "tomorrow," not after a merge. :-) 

If the 'Split()' call results in a single element, there is no prefix, and so the children shouldn't be examined. I'll fix that tomorrow and push another update.

File: src/Native/LibTorchSharp/THSNormalization.cpp [hunk 5]
```diff
@@ -316,6 +334,15 @@ Tensor THSNN_BatchNorm3d_get_var(const NNModule module)
     return nullptr;
 }
 
+Tensor THSNN_BatchNorm3d_get_batches(const NNModule module)
+{
+    CATCH(
+        auto v = (*module)->as<torch::nn::BatchNorm3d>()->num_batches_tracked;
+    return v.defined() ? ResultTensor(v) : nullptr;
```
Comment (lostmsu) [inline src/Native/LibTorchSharp/THSNormalization.cpp:341]: NIT :-D

File: src/TorchSharp/NN/Module.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs]: Okay, I've updated with what I think is a correct implementation.

File: src/TorchSharp/NN/Module.cs [hunk 6]
```diff
@@ -512,14 +515,40 @@ public virtual bool has_parameter(string target)
...
                     }
-                    foreach (var child in named_children().Where(nc => target.StartsWith(nc.name))) {
-                        var prefix = child.name + ".";
-                        if (child.module.has_parameter(target.Remove(0, prefix.Length)))
-                            return true;
+
+                    var splits = target.Split('.');
+                    if (splits.Length > 1) {
+                        foreach (var child in named_children().Where(nc => nc.name == splits[0])) {
+                            if (child.module.has_parameter(target.Remove(0, splits[0].Length + 1)))
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:522]: NIT: `target[(splits[0].Length + 2)..]`

File: src/TorchSharp/NN/Module.cs [hunk 7]
```diff
@@ -512,14 +515,40 @@ public virtual bool has_parameter(string target)
...
+                    var splits = target.Split('.');
+                    if (splits.Length > 1) {
+                        foreach (var child in named_children().Where(nc => nc.name == splits[0])) {
+                            var p = child.module.get_buffer(target.Remove(0, splits[0].Length + 1));
+                            if (p is not null)
+                                return p;
+                        }
+                    }
+                    return null;
+                }
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:550]: NIT: this repeats a bit. Perhaps make a `T get_named<T>(name, internal_collection, getter)`?

**Review by lostmsu — APPROVED** [review]
I'd squash.

**Pull request merged**